openmc/volume.py

```diff
@@ -0,0 +1,280 @@
+from collections import OrderedDict
+from collections.abc import Iterable, Mapping
+from numbers import Real, Integral
+from xml.etree import ElementTree as ET
+import warnings
+
+import numpy as np
+import pandas as pd
+import h5py
+from uncertainties import ufloat
+
+import openmc
+import openmc.checkvalue as cv
+
+_VERSION_VOLUME = 1
+
+
+class VolumeCalculation(object):
+    """Stochastic volume calculation specifications and results.
+
+    Parameters
+    ----------
+    domains : Iterable of openmc.Cell, openmc.Material, or openmc.Universe
+        Domains to find volumes of
+    samples : int
+        Number of samples used to generate volume estimates
+    lower_left : Iterable of float
+        Lower-left coordinates of bounding box used to sample points. If this
+        argument is not supplied, an attempt is made to automatically determine
+        a bounding box.
+    upper_right : Iterable of float
+        Upper-right coordinates of bounding box used to sample points. If this
+        argument is not supplied, an attempt is made to automatically determine
+        a bounding box.
+
+    Attributes
+    ----------
+    ids : Iterable of int
+        IDs of domains to find volumes of
+    domain_type : {'cell', 'material', 'universe'}
+        Type of each domain
+    samples : int
+        Number of samples used to generate volume estimates
+    lower_left : Iterable of float
+        Lower-left coordinates of bounding box used to sample points
+    upper_right : Iterable of float
+        Upper-right coordinates of bounding box used to sample points
+    atoms : dict
+        Dictionary mapping unique IDs of domains to a mapping of nuclides to
+        total number of atoms for each nuclide present in the domain. For
+        example, {10: {'U235': 1.0e22, 'U238': 5.0e22, ...}}.
+    atoms_dataframe : pandas.DataFrame
+        DataFrame showing the estimated number of atoms for each nuclide present
+        in each domain specified.
+    volumes : dict
+        Dictionary mapping unique IDs of domains to estimated volumes in cm^3.
+
+    """
+    def __init__(self, domains, samples, lower_left=None,
+                 upper_right=None):
+        self._atoms = {}
+        self._volumes = {}
+
+        cv.check_type('domains', domains, Iterable,
+                      (openmc.Cell, openmc.Material, openmc.Universe))
+        if isinstance(domains[0], openmc.Cell):
+            self._domain_type = 'cell'
+        elif isinstance(domains[0], openmc.Material):
+            self._domain_type = 'material'
+        elif isinstance(domains[0], openmc.Universe):
+            self._domain_type = 'universe'
+        self.ids = [d.id for d in domains]
+
+        self.samples = samples
+
+        if lower_left is not None:
+            if upper_right is None:
```

Comment:
Should this block be moved to the `else` clause below? The exception says that both coordinates should be specified, and this check will trigger if `lower_left` is specified but `lower_right` is not, which is actually a valid code path according to this class's docstring.